### ISSUE TITLE
General wisdom

### ISSUE THREAD

**Issue by mtewes** [mtewes/tenbilac#4]
Regarding the code structure, should the Normer objects be considered as attributes of the Tenbilac object ?
In any case one has to be careful not to mix Normers constructed with different training sets!

**Comment by mtewes** [comment]
The error functions accept masked arrays. But their "mean" accross training cases should be weighted according to the number of maked realizations for each case. Maybe simply switch to sums ?

**Comment by mtewes** [comment]
> The error functions accept masked arrays. But their "mean" accross training cases should be weighted according to the number of maked realizations for each case. Maybe simply switch to sums ?

No, make a chi2 - like metric, so that we expect 1.0

**Comment by mtewes** [comment]
The default  act is now Tanh, and also Sig is made even by rescaling.
Do not forget to make Normer mode "-11" the default as well !
(Check which combi is faster  / better ! -> make these options !)

**Comment by mtewes** [comment]
Viz the network parameters, and add a regularization!

**Comment by mtewes** [comment]
Insert a new layer in cascade, preset it to identity. Train only this new layer.

**Comment by mtewes** [comment]
The "chi2"-like MSRB is implemented. Seems great.
![screen shot 2015-07-13 at 16 52 03](https://cloud.githubusercontent.com/assets/1928653/8652563/9536d420-297f-11e5-94b0-615f71aff905.png)

**Comment by mtewes** [comment]
Viz of params:
![screen shot 2015-07-13 at 17 15 05](https://cloud.githubusercontent.com/assets/1928653/8653079/c050ec92-2982-11e5-812f-dcc7c5399199.png)

**Comment by mtewes** [comment]
- Add time to callback
- implement mini-batch !!!
- profile...

**Comment by mtewes** [comment]
> Add time to callback
> implement mini-batch !!!

done and done

profiling: not yet

**Comment by mtewes** [comment]
automatic minibatch size could be x times self.nparams()

**Comment by mtewes** [comment]
in the wrapper, how to add possibility to restart training. Maybe just by copying existing params if compatible.

**Comment by mtewes** [comment]
Add cross-validation set to Training object.

**Comment by mtewes** [comment]
Max square relative bias!

**Comment by mtewes** [comment]
cross-validation done. Now implementing restart in wrapper, to see how CV looks for MegaLUT.

**Comment by mtewes** [comment]
A stopping criterion within a mini-batch could be the CV/minibatch ratio getting to high.

**Comment by mtewes** [comment]
Todo: 
- color code to paramsplot
- errfct terms plots (as func of input) + carry around names for feature indices
- wrapper: find way to keep train logs when reuse=True.

**Comment by mtewes** [comment]
Done with color code + carry around names. 
Now implementing the errfct terms plot!

**Comment by mtewes** [comment]
Managed to get once a fantastic size-demo-"failure" with MSRB, with a good cost function but a very unsmooth curve. Might indicate that a mixture of MSRB and MSE is actually preferable ?

**Comment by mtewes** [comment]
Build a MSRE, so a weighted MSE

**Comment by mtewes** [comment]
Size-demo thoughts: on the dependance of the training distributions:

- distrib of true parameters used for training: does matter for the estimator as
  - one line = act as a regularization, limited in flexiblity ?
  - if we use *mean* square something, the distribs change "weights" that enter into this mean.

- test distribs that query the network: do not matter for the aspect of the plot. Can matter if you "fit" something like a line to a plot, as it weights different regions of true parameters space differently.
